tornado/locks.py

```diff
@@ -79,3 +79,51 @@ def _garbage_collect(self):
             self._timeouts = 0
             self._waiters = collections.deque(
                 w for w in self._waiters if not w.done())
+
+_true_future = Future()
+_true_future.set_result(True)
+
+
+class Event(object):
+    """An event blocks coroutines until its internal flag is set to True.
+
+    Similar to `threading.Event`.
+    """
+    def __init__(self):
+        self.io_loop = ioloop.IOLoop.current()
+        self._condition = Condition()
+        self._flag = False
+
+    def __str__(self):
+        return '<%s %s>' % (
+            self.__class__.__name__, 'set' if self._flag else 'clear')
+
+    def is_set(self):
+        """Return ``True`` if the internal flag is true."""
+        return self._flag
+
+    def set(self):
+        """Set the internal flag to ``True``. All waiters are awakened.
+
+        Calling `.wait` once the flag is set will not block.
+        """
+        self._flag = True
+        self._condition.notify_all()
+
+    def clear(self):
+        """Reset the internal flag to ``False``.
+        
+        Calls to `.wait` will block until `.set` is called.
+        """
+        self._flag = False
+
+    def wait(self, deadline=None):
+        """Block until the internal flag is true.
+
+        Returns a `.Future` that resolves ``True`` if the condition is notified,
+        or ``False`` after a timeout.
```

Comment:
Hmm. I think I would part ways with `threading.Event` here and raise an exception on timeout. The reason is that condition variables by their nature and by tradition allow for spurious wakeups, but Events want to guarantee that if wait() returns normally the flag has been set (I suspect threading.Event would have used a timeout instead of a return value if they were not bound by backwards-compatibility with py26, where there was no indication that a timeout had occurred).

If timeout raises instead of returns, then the timeout behavior is equivalent to gen.with_timeout. In fact, the whole class can collapse into a simple wrapper around a Future. The only attribute of an Event is a Future, and we define is_set = self._future.done(); set = self._future.set_result(None); wait = `return self._future if deadline is None else gen.with_timeout(timeout, self._future)`